File: src/main/java/cs/java/collections/CSList.java
package cs.java.collections;

import java.util.List;

import androidx.annotation.Nullable;
import cs.java.lang.CSValues;

public interface CSList<T> extends java.util.List<T>, CSValues<T> {

    @Nullable T at(int index);

    int delete(T item);

    T first();

    T second();

    int index(T item);

    boolean getHasItems();

    @Override
    @Deprecated
    int indexOf(Object arg0);

    boolean isLast(T item);

    boolean isLastIndex(int index);

    @Nullable T last();

    @Nullable T previous();

    int count();

    int lastIndex();

    CSList<T> rangeFrom(int fromIndex);

    CSList<T> range(int fromIndex, int toIndex);

    boolean remove(Object arg0);

    T removeLast();

    CSList<T> removeAll();

    T put(T item);

    T add(T item, int index);

    T set(T item, int index);

    CSList<T> append(T... item);

    CSList<T> append(List<T> items);

    CSList<T> insert(int index, T item);

    CSList<T> reload(List<T> values);

    boolean has(T bow);

    int length();

    CSList<T> reverse();
}

File: library/renetik-android/src/main/java/renetik/java/collections/CSList.kt
package renetik.java.collections

import renetik.java.lang.CSValues

fun <T> list(): CSList<T> = CSListImpl<T>()

fun <T> list(vararg items: T): CSList<T> = list<T>().putAll(*items)

fun <T> list(items: Iterable<T>): CSList<T> = list<T>().putAll(items)

interface CSList<T> : MutableList<T>, List<T>, CSValues<T> {

    val hasItems: Boolean

    fun at(index: Int): T?
    fun first(): T?
    fun second(): T?
    fun last(): T?
    fun previousLast(): T?

    fun isLast(item: T): Boolean
    fun isLastIndex(index: Int): Boolean

    fun lastIndex(): Int

    fun rangeFrom(fromIndex: Int): CSList<T>
    fun range(fromIndex: Int, toIndex: Int): CSList<T>

    fun put(item: T): T
    fun put(item: T, index: Int): T
    fun putAll(vararg items: T): CSList<T>
    fun putAll(items: Iterable<T>): CSList<T>
    fun replace(item: T, index: Int): T
    fun reload(values: Iterable<T>): CSList<T>

    fun reverse(): CSList<T>

    fun delete(index: Int): T?
    fun delete(item: T): Int
    fun deleteLast(): T?
    fun deleteAll(): CSList<T>

}
